File: src/decoders/authentication-qr.ts
import type { AuthenticationQR } from "~/models";
import { decodeAccountKindFromPath } from "./account-kind";

export const decodeAuthenticationQr = (qr: any): AuthenticationQR => {
  const url = qr.url as string;

  const kindIndex: number = url.lastIndexOf("/") + 1;
  const base = url.substring(0, kindIndex - 1);
  const kindPath = url.substring(kindIndex);

  return {
    kind: decodeAccountKindFromPath(kindPath),
    url: base,
    token: qr.jeton,
    username: qr.login
  };
};

File: library/src/commonMain/kotlin/decoders/authentication-qr.kt
package decoders

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import models.AuthenticationQR

fun decodeAuthenticationQr (qr: JsonObject): AuthenticationQR {
    val url = qr["url"]!!.jsonPrimitive.content

    val kindIndex = url.lastIndexOf("/") + 1
    val base = url.substring(0, kindIndex - 1)
    val kindPath = url.substring(kindIndex)

    return AuthenticationQR(
        kind = decodeAccountKindFromPath(kindPath),
        url = base,
        token = qr["jeton"]!!.jsonPrimitive.content,
        username = qr["login"]!!.jsonPrimitive.content
    )
}